### PR TITLE
Large changes to repeated field handling

### PR DESCRIPTION
**Pull request by jskeet**
Moved all the code for repeated field handling out of Coded*Stream (which are now much more "primitive"-oriented) and into RepeatedField.

This PR includes quite a lot of tidying up, extra tests, more XML documentation as well.
Unfortunately it's relatively tricky to split into C++, generated code and runtime code at this point.

I'm confident in the tests, however - I've got good coverage, and there's now a lot less code *to* test, albeit it's slightly trickier code.

@anandolee - I know @jtattermusch is out for another week. Let me know if you can think of a way of making this set of changes simpler to organize. (I don't know if there's a way I could rebase but change how it's all committed into separate commits etc...)

### CODE REVIEW

File: csharp/src/ProtocolBuffers.Test/CodedInputStreamTest.cs [hunk 1]
```diff
@@ -232,66 +230,26 @@ public void DecodeZigZag64()
...
             Assert.AreEqual(unchecked((long) 0x8000000000000000L), CodedInputStream.DecodeZigZag64(0xFFFFFFFFFFFFFFFFL));
         }
-        /*
+        
         [Test]
-        public void ReadWholeMessage()
+        public void ReadWholeMessage_VaryingBlockSizes()
         {
-            TestAllTypes message = TestUtil.GetAllSet();
+            TestAllTypes message = GeneratedMessageTest.GetSampleMessage();
```
Comment (anandolee) [inline csharp/src/ProtocolBuffers.Test/CodedInputStreamTest.cs]: Are you going to remove TestUtil? I'd prefer to keep it for generated code tests.

Comment (jskeet) [inline csharp/src/ProtocolBuffers.Test/CodedInputStreamTest.cs]: Most of the code within it was becoming useless - it ended up really only being useful for getting a sample message. I can move that back in if you think it's better there though. (Probably not today or tomorrow - I'm in Zurich at the moment.)

**Comment by anandolee** [comment]
Size of this PR is OK for me.

File: csharp/src/ProtocolBuffers/CodedOutputStream.cs [hunk 2]
```diff
@@ -534,178 +443,13 @@ public void WriteRawTag(byte b1, byte b2, byte b3, byte b4, byte b5)
...
         #endregion
 
-        #region Write packed array members
-        // TODO(jonskeet): A lot of these are really inefficient, due to method group conversions. Fix!
-        // (Alternatively, add extension methods to RepeatedField, accepting the Write* methods via delegates too.)
-        public void WritePackedBoolArray(RepeatedField<bool> list)
-        {
-            uint size = (uint)list.Count;
-            WriteRawVarint32(size);
-            foreach (var value in list)
```
Comment (jtattermusch) [inline csharp/src/ProtocolBuffers/CodedOutputStream.cs]: We've found out that foreach can be slow, right? for(int i = 0; ....) might be better here (and below).

File: csharp/src/ProtocolBuffers.Test/CodedInputStreamTest.cs (hunk 1 above)
Comment (jtattermusch) [inline csharp/src/ProtocolBuffers.Test/CodedInputStreamTest.cs]: I am not huge fan of using GeneratedMessageTest.GetSampleMessage() from CodedInputStreamTest. There should be a separate class if we need to call GetSampleMessage from multiple classes. The name could be more specific than just TestUtil.

File: csharp/src/ProtocolBuffers.Test/SampleEnum.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+namespace Google.Protobuf
+{
+    // Just a sample enum with positive and negative values to be used in tests.
+    internal enum SampleEnum
+    {
+        NegativeValue = -2,
+        None = 0,
+        PositiveValue = 3
+    }
+}
```
Comment (jtattermusch) [inline csharp/src/ProtocolBuffers.Test/SampleEnum.cs]: add EOL

**Comment by jtattermusch** [comment]
I like that this PR simplifies the generated code quite a bit.

File: csharp/src/AddressBook/Addressbook.cs [hunk 4]
```diff
@@ -241,7 +236,7 @@ public sealed partial class Person : pb::IMessage<Person> {
             break;
           }
           case 34: {
-            input.ReadMessageArray(phone_, global::Google.Protobuf.Examples.AddressBook.Person.Types.PhoneNumber.Parser);
+            phone_.AddEntriesFrom(input, _repeated_phone_codec);
```
Comment (jtattermusch) [inline csharp/src/AddressBook/Addressbook.cs]: is there a reason why phone_ is separate from the _repeated_phone_codec?   the phone_ field will only ever use that specific codec, right?
the mehod signature   AddEntriesFrom(codedInputStream, codec) looks a bit odd - perhaps we could make it just phone_.AddEntriesFrom(input) and phone would know which codec to use? It is completely fine to do this in the future (if ever), I actually don't feel strong about this.

File: csharp/src/ProtocolBuffers/CodedInputStream.cs [hunk 5]
```diff
@@ -435,24 +520,16 @@ public long ReadSInt64()
...
 
-        private bool BeginArray(uint fieldTag, out bool isPacked, out int oldLimit)
+        /// <summary>
+        /// Reads a length for length-delimited data.
+        /// </summary>
+        /// <remarks>
+        /// This is internally just reading a varint, but this method exists
+        /// to make the calling code clearer.
+        /// </remarks>
+        public int ReadLength()
```
Comment (jtattermusch) [inline csharp/src/ProtocolBuffers/CodedInputStream.cs]: I like having an extra method just for reading length, it really improves the readability!

File: csharp/src/ProtocolBuffers/CodedInputStream.cs [hunk 6]
```diff
@@ -793,12 +608,12 @@ private uint SlowReadRawVarint32()
...
 
         /// <summary>
-        /// Read a raw Varint from the stream.  If larger than 32 bits, discard the upper bits.
+        /// Reads a raw Varint from the stream.  If larger than 32 bits, discard the upper bits.
         /// This method is optimised for the case where we've got lots of data in the buffer.
         /// That means we can check the size just once, then just read directly from the buffer
         /// without constant rechecking of the buffer length.
         /// </summary>
-        public uint ReadRawVarint32()
+        internal uint ReadRawVarint32()
```
Comment (jtattermusch) [inline csharp/src/ProtocolBuffers/CodedInputStream.cs]: Will external code still have a way to use the Raw methods? They are probably not terribly useful but having access to the low level encoding routines might come handy from time to time if someone tries to hack something non-standard (or perhaps in tests).

Comment (jtattermusch) [inline csharp/src/ProtocolBuffers/CodedInputStream.cs]: Ah, the public methods exist. please ignore this comment.

File: csharp/src/ProtocolBuffers/FieldCodec.cs [hunk 7]
```diff
@@ -145,28 +146,68 @@ static FieldCodec()
...
+            this.sizeCalculator = _ => fixedSize;
+            this.fixedSize = fixedSize;
+            this.tag = tag;
+            tagSize = CodedOutputStream.ComputeRawVarint32Size(tag);
+        }
+
+        /// <summary>
+        /// Returns the size calculator for just a value.
+        /// </summary>
+        internal Func<T, int> ValueSizeCalculator { get { return sizeCalculator; } }
```
Comment (jtattermusch) [inline csharp/src/ProtocolBuffers/FieldCodec.cs]: What does it buy us to have a FieldCodec<T> with properties of Func/Action types? We could also have a IFieldCodec interface with methods like Write(), Read(), CalculateSize() that does the same thing. My suspicion is that the interface approach might be a bit faster as well.

**Comment by jtattermusch** [comment]
Besides the comments above, LGTM.  After briefly discussing the comments, I'll be happy to merge.

File: csharp/src/ProtocolBuffers.Test/CodedInputStreamTest.cs (hunk 1 above)
Comment (jskeet) [inline csharp/src/ProtocolBuffers.Test/CodedInputStreamTest.cs]: Okay, that's fine by me. Given the number of PRs in the "train" now, do you mind if I do that after we've merged the PRs? Doing a lot of rebasing now for something that can easily be done later (probably in a complete separate PR) feels error-prone, at least with my level of git competence.

File: csharp/src/ProtocolBuffers/CodedOutputStream.cs (hunk 2 above)
Comment (jskeet) [inline csharp/src/ProtocolBuffers/CodedOutputStream.cs]: This is deleted code anyway :)

File: csharp/src/AddressBook/Addressbook.cs (hunk 4 above)
Comment (jskeet) [inline csharp/src/AddressBook/Addressbook.cs]: I wondered about that, but note that the codec is static whereas the list is a per-instance field. Making it part of the list would mean having the same value in every list, and potentially creating a new codec each time.

File: csharp/src/ProtocolBuffers/CodedInputStream.cs (hunk 5 above)
Comment (jskeet) [inline csharp/src/ProtocolBuffers/CodedInputStream.cs]: Me too - I was fed up with all those casts :)

File: csharp/src/ProtocolBuffers/FieldCodec.cs (hunk 7 above)
Comment (jskeet) [inline csharp/src/ProtocolBuffers/FieldCodec.cs]: We need to call these delegates repeatedly when serializing/deserializing arrays. This reduces the number of levels of indirection. We could potentially investigate it later though.

File: csharp/src/ProtocolBuffers.Test/SampleEnum.cs (hunk 3 above)
Comment (jskeet) [inline csharp/src/ProtocolBuffers.Test/SampleEnum.cs]: As per earlier comment, I'd like to punt on this to a future PR if possible, given that it's trivial.

**Comment by jskeet** [comment]
Given the awkwardness of the time zones, I'm going to merge this PR and the next one, create a small "tidying-up" for the two minor comments above - which hopefully can be merged overnight. (I'll rebase the reflection changes one onto that.)

**Comment by jtattermusch** [comment]
Your comments make sense to me, thanks!

**Pull request merged**